File: src/LeetCode/BinaryTree/JZ32.java
package LeetCode.BinaryTree;

import datastructure.BinaryTree.TreeNode;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

public class JZ32 {
    /**
     * 从上到下按层打印二叉树，同一层的节点按从左到右的顺序打印，每一层打印到一行。
     */
    class Solution {
        public List<List<Integer>> levelOrder(TreeNode root) {
            List<List<Integer>> res = new LinkedList<>();
            if (root == null) {
                return res;
            }

            Deque<TreeNode> queue = new LinkedList<>();
            queue.addLast(root);
            while (!queue.isEmpty()) {
                List<Integer> subList = new LinkedList<>();
                int size = queue.size();
                for (int i = 0; i < size; i++) {
                    TreeNode temp = queue.pollFirst();
                    subList.add(temp.val);
                    if (temp.left != null) {
                        queue.addLast(temp.left);
                    }
                    if (temp.right != null) {
                        queue.addLast(temp.right);
                    }
                }
                res.add(subList);
            }
            return res;
        }
    }

    public static void main(String[] args) {
        TreeNode head = new TreeNode(3);
        TreeNode nine = new TreeNode(9);
        TreeNode twenty = new TreeNode(20);
        TreeNode fifteen = new TreeNode(15);
        TreeNode seven = new TreeNode(7);
        head.left = nine;
        head.right = twenty;
        twenty.left = fifteen;
        twenty.right = seven;
        new JZ32().new Solution().levelOrder(head);
    }
}

File: LeetcodePY/binaryTree/JZ32.py
#从上到下按层打印二叉树，同一层的节点按从左到右的顺序打印，每一层打印到一行。
# Definition for a binary tree node.
import collections


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def levelOrder(self, root: TreeNode):
        if not root: return []
        res,queue = [], collections.deque()
        queue.append(root)
        while queue:
            temp=[]
            for _ in range(len(queue)):
                node = queue.popleft()
                temp.append(node.val)
                if node.left:queue.append(node.left)
                if node.right: queue.append(node.right)
            res.append(temp)
        return res